Coercion from datetime at midnight to date is incompatible with JSON schema

### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

In Pydantic V2 we allowed date to be constructed from a timestamp string if the timestamp is exactly at midnight.

In https://github.com/pydantic/pydantic/issues/7039#issuecomment-1671859377 @MarkusSintonen makes some compelling arguments why this was a mistake:
- It's not accepted in JSON schema for "date" format
- It's not accepted in Python's `date.fromisoformat`
- It wasn't accepted in V1

On a similar vein, we also coerce `datetime` to a `date` if the datetime is midnight.

I think we should consider classifying these coercions as a bug and reverting before it is too late.

If users want this behaviour, they can implement a `"before"` validator which does these coercions with `Annotated`. 

### Example Code

```Python
>>> from pydantic import TypeAdapter
>>> from datetime import date, datetime

>>> TypeAdapter(date).validate_python("2020-01-01T00:00:00")
datetime.date(2020, 1, 1)

>>> TypeAdapter(date).validate_python(datetime(2021, 1, 1))
datetime.date(2021, 1, 1)
```


### Python, Pydantic & OS Version

```Text
pydantic version: 2.1.1
        pydantic-core version: 2.4.0
          pydantic-core build: profile=release pgo=true mimalloc=true
                 install path: /home/david/.virtualenvs/pydantic-prod/lib/python3.11/site-packages/pydantic
               python version: 3.11.4 (main, Jun  9 2023, 07:59:55) [GCC 12.3.0]
                     platform: Linux-5.15.90.1-microsoft-standard-WSL2-x86_64-with-glibc2.37
     optional deps. installed: ['email-validator', 'typing-extensions']
```


Selected Assignee: @Kludex

> just as .0 floats are allowed as input to int fields
> The rule we're trying to stick to in V2 is "in lax mode, coerce if there's one obvious equivalent in the field type, and no data is lost".

Comparing numeric value handling with datetime parsing rules doesnt make much sense to me. Why do you think its a good idea to convert ISO datetimes (at midnights) with TZs into date instances? V2s strange ISO datetime -> date coercion loses now information that you said should be avoiding in "lax" mode. This really feels like a bug rather than a good feature:

```python
class A(BaseModel):
    a: date | datetime

# Fails with V2, works in V1
# AssertionError: assert datetime.date(2023, 1, 1) == datetime.datetime(2023, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200)))
assert A(a="2023-01-01T00:00:00+02:00").a == datetime(2023, 1, 1, tzinfo=timezone(timedelta(hours=2)))
```

> The other problem is that fastapi doesn't use validate_json, hence users are effectively forced to use lax mode with fastapi 
> As a workaround we could add a config flag to prevent all datetimes as inputs to a date field, default off. That would fix the problem for users that want that, without changing current behaviour.

I dont think that is a valid point or workaround as there is much more of parameter parsing in FastAPI that is not anything related to json. See trivial example below which is now broken in V2. Adding yet more parameters to configs doesnt feel right they are not really usable in many contexts. This is especially bad if the lossy conversion is on by default. It would be horrible usage ergonomics if you have to always remember to opt-out from the current behaviour all over the place.

```python
def test_fastapi():
    async def handler(v: date):
        return {"v": v}

    app = FastAPI()
    app.router.add_api_route("/foo", handler)

    # API accepts date https://json-schema.org/understanding-json-schema/reference/string.html#dates-and-times
    assert app.openapi()["paths"]["/foo"]["get"]["parameters"][0]["schema"] == {"format": "date", "title": "V", "type": "string"}

    client = TestClient(app)
    res = client.get("/foo", params={"v": "2023-01-01T00:00:00+02:00"})
    # In Pydantic V2 this test fails as it gives an incorrect successful response...
    # AssertionError: Got: {'v': '2023-01-01'}
    # assert 200 == 422
    assert res.status_code == 422, f"Got: {res.json()}"
```

Neither its an issue in FastAPI.

--- a/pydantic/main.py
+++ b/pydantic/main.py
@@ -531,6 +531,28 @@ class BaseModel(metaclass=_model_construction.ModelMetaclass):
         return cls.__pydantic_validator__.validate_json(json_data, strict=strict, context=context)
 
     @classmethod
+    def model_validate_strings(
+        cls: type[Model],
+        obj: Any,
+        *,
+        strict: bool | None = None,
+        context: dict[str, Any] | None = None,
+    ) -> Model:
+        """Validate the given object contains string data against the Pydantic model.
+
+        Args:
+            obj: The object contains string data to validate.
+            strict: Whether to enforce types strictly.
+            context: Extra variables to pass to the validator.
+
+        Returns:
+            The validated Pydantic model.
+        """
+        # `__tracebackhide__` tells pytest and some other tools to omit this function from tracebacks
+        __tracebackhide__ = True
+        return cls.__pydantic_validator__.validate_strings(obj, strict=strict, context=context)
+
+    @classmethod
     def __get_pydantic_core_schema__(
         cls, __source: type[BaseModel], __handler: _annotated_handlers.GetCoreSchemaHandler
     ) -> CoreSchema:

--- a/pydantic/type_adapter.py
+++ b/pydantic/type_adapter.py
@@ -221,6 +221,19 @@ class TypeAdapter(Generic[T]):
         """
         return self.validator.validate_json(__data, strict=strict, context=context)
 
+    def validate_strings(self, __obj: Any, *, strict: bool | None = None, context: dict[str, Any] | None = None) -> T:
+        """Validate object contains string data against the model.
+
+        Args:
+            __obj: The object contains string data to validate.
+            strict: Whether to strictly check types.
+            context: Additional context to use during validation.
+
+        Returns:
+            The validated object.
+        """
+        return self.validator.validate_strings(__obj, strict=strict, context=context)
+
     def get_default_value(self, *, strict: bool | None = None, context: dict[str, Any] | None = None) -> Some[T] | None:
         """Get the default value for the wrapped type.
 

--- a/tests/test_dataclasses.py
+++ b/tests/test_dataclasses.py
@@ -6,7 +6,7 @@ import sys
 import traceback
 from collections.abc import Hashable
 from dataclasses import InitVar
-from datetime import datetime
+from datetime import date, datetime
 from pathlib import Path
 from typing import Any, Callable, ClassVar, Dict, FrozenSet, Generic, List, Optional, Set, TypeVar, Union
 
@@ -2593,3 +2593,14 @@ def test_alias_with_dashes():
 
     obj = Foo(**{'some-var': 'some_value'})
     assert obj.some_var == 'some_value'
+
+
+def test_validate_strings():
+    @pydantic.dataclasses.dataclass
+    class Nested:
+        d: date
+
+    class Model(BaseModel):
+        n: Nested
+
+    assert Model.model_validate_strings({'n': {'d': '2017-01-01'}}).n.d == date(2017, 1, 1)

--- a/tests/test_main.py
+++ b/tests/test_main.py
@@ -5,6 +5,7 @@ import sys
 from collections import defaultdict
 from copy import deepcopy
 from dataclasses import dataclass
+from datetime import date, datetime
 from enum import Enum
 from typing import (
     Any,
@@ -2541,6 +2542,53 @@ def test_validate_python_from_attributes() -> None:
 
     res = ModelFromAttributesFalse.model_validate(UnrelatedClass(), from_attributes=True)
     assert res == ModelFromAttributesFalse(x=1)
+
+
+@pytest.mark.parametrize(
+    'field_type,input_value,expected,raises_match,strict',
+    [
+        (bool, 'true', True, None, False),
+        (bool, 'true', True, None, True),
+        (bool, 'false', False, None, False),
+        (bool, 'e', ValidationError, 'type=bool_parsing', False),
+        (int, '1', 1, None, False),
+        (int, '1', 1, None, True),
+        (int, 'xxx', ValidationError, 'type=int_parsing', True),
+        (float, '1.1', 1.1, None, False),
+        (float, '1.10', 1.1, None, False),
+        (float, '1.1', 1.1, None, True),
+        (float, '1.10', 1.1, None, True),
+        (date, '2017-01-01', date(2017, 1, 1), None, False),
+        (date, '2017-01-01', date(2017, 1, 1), None, True),
+        (date, '2017-01-01T12:13:14.567', ValidationError, 'type=date_from_datetime_inexact', False),
+        (date, '2017-01-01T12:13:14.567', ValidationError, 'type=date_parsing', True),
+        (date, '2017-01-01T00:00:00', date(2017, 1, 1), None, False),
+        (date, '2017-01-01T00:00:00', ValidationError, 'type=date_parsing', True),
+        (datetime, '2017-01-01T12:13:14.567', datetime(2017, 1, 1, 12, 13, 14, 567_000), None, False),
+        (datetime, '2017-01-01T12:13:14.567', datetime(2017, 1, 1, 12, 13, 14, 567_000), None, True),
+    ],
+    ids=repr,
+)
+def test_model_validate_strings(field_type, input_value, expected, raises_match, strict):
+    class Model(BaseModel):
+        x: field_type
+
+    if raises_match is not None:
+        with pytest.raises(expected, match=raises_match):
+            Model.model_validate_strings({'x': input_value}, strict=strict)
+    else:
+        Model.model_validate_strings({'x': input_value}, strict=strict).x == expected
+
+
+@pytest.mark.parametrize('strict', [True, False])
+def test_model_validate_strings_dict(strict):
+    class Model(BaseModel):
+        x: Dict[int, date]
+
+    assert Model.model_validate_strings({'x': {'1': '2017-01-01', '2': '2017-01-02'}}, strict=strict).x == {
+        1: date(2017, 1, 1),
+        2: date(2017, 1, 2),
+    }
 
 
 def test_model_signature_annotated() -> None:

--- a/tests/test_main.py
+++ b/tests/test_main.py
@@ -2577,7 +2577,7 @@ def test_model_validate_strings(field_type, input_value, expected, raises_match,
         with pytest.raises(expected, match=raises_match):
             Model.model_validate_strings({'x': input_value}, strict=strict)
     else:
-        Model.model_validate_strings({'x': input_value}, strict=strict).x == expected
+        assert Model.model_validate_strings({'x': input_value}, strict=strict).x == expected
 
 
 @pytest.mark.parametrize('strict', [True, False])

--- a/tests/test_root_model.py
+++ b/tests/test_root_model.py
@@ -1,4 +1,5 @@
 import pickle
+from datetime import date, datetime
 from typing import Any, Dict, List, Optional, Union
 
 import pytest
@@ -616,3 +617,38 @@ def test_copy_preserves_equality():
 
     deepcopied = model.__deepcopy__()
     assert model == deepcopied
+
+
+@pytest.mark.parametrize(
+    'root_type,input_value,expected,raises_match,strict',
+    [
+        (bool, 'true', True, None, False),
+        (bool, 'true', True, None, True),
+        (bool, 'false', False, None, False),
+        (bool, 'e', ValidationError, 'type=bool_parsing', False),
+        (int, '1', 1, None, False),
+        (int, '1', 1, None, True),
+        (int, 'xxx', ValidationError, 'type=int_parsing', True),
+        (float, '1.1', 1.1, None, False),
+        (float, '1.10', 1.1, None, False),
+        (float, '1.1', 1.1, None, True),
+        (float, '1.10', 1.1, None, True),
+        (date, '2017-01-01', date(2017, 1, 1), None, False),
+        (date, '2017-01-01', date(2017, 1, 1), None, True),
+        (date, '2017-01-01T12:13:14.567', ValidationError, 'type=date_from_datetime_inexact', False),
+        (date, '2017-01-01T12:13:14.567', ValidationError, 'type=date_parsing', True),
+        (date, '2017-01-01T00:00:00', date(2017, 1, 1), None, False),
+        (date, '2017-01-01T00:00:00', ValidationError, 'type=date_parsing', True),
+        (datetime, '2017-01-01T12:13:14.567', datetime(2017, 1, 1, 12, 13, 14, 567_000), None, False),
+        (datetime, '2017-01-01T12:13:14.567', datetime(2017, 1, 1, 12, 13, 14, 567_000), None, True),
+    ],
+    ids=repr,
+)
+def test_model_validate_strings(root_type, input_value, expected, raises_match, strict):
+    Model = RootModel[root_type]
+
+    if raises_match is not None:
+        with pytest.raises(expected, match=raises_match):
+            Model.model_validate_strings(input_value, strict=strict)
+    else:
+        assert Model.model_validate_strings(input_value, strict=strict).root == expected

--- a/tests/test_type_adapter.py
+++ b/tests/test_type_adapter.py
@@ -295,12 +295,12 @@ def test_validate_python_from_attributes() -> None:
     ids=repr,
 )
 def test_validate_strings(field_type, input_value, expected, raises_match, strict):
+    ta = TypeAdapter(field_type)
     if raises_match is not None:
-        print(TypeAdapter(field_type).core_schema)
         with pytest.raises(expected, match=raises_match):
-            TypeAdapter(field_type).validate_strings(input_value, strict=strict)
+            ta.validate_strings(input_value, strict=strict)
     else:
-        TypeAdapter(field_type).validate_strings(input_value, strict=strict) == expected
+        assert ta.validate_strings(input_value, strict=strict) == expected
 
 
 @pytest.mark.parametrize('strict', [True, False])

--- a/tests/test_type_adapter.py
+++ b/tests/test_type_adapter.py
@@ -1,6 +1,7 @@
 import json
 import sys
 from dataclasses import dataclass
+from datetime import date, datetime
 from typing import Any, Dict, ForwardRef, Generic, List, NamedTuple, Tuple, TypeVar, Union
 
 import pytest
@@ -266,3 +267,45 @@ def test_validate_python_from_attributes() -> None:
 
     res = ta.validate_python(UnrelatedClass(), from_attributes=True)
     assert res == ModelFromAttributesFalse(x=1)
+
+
+@pytest.mark.parametrize(
+    'field_type,input_value,expected,raises_match,strict',
+    [
+        (bool, 'true', True, None, False),
+        (bool, 'true', True, None, True),
+        (bool, 'false', False, None, False),
+        (bool, 'e', ValidationError, 'type=bool_parsing', False),
+        (int, '1', 1, None, False),
+        (int, '1', 1, None, True),
+        (int, 'xxx', ValidationError, 'type=int_parsing', True),
+        (float, '1.1', 1.1, None, False),
+        (float, '1.10', 1.1, None, False),
+        (float, '1.1', 1.1, None, True),
+        (float, '1.10', 1.1, None, True),
+        (date, '2017-01-01', date(2017, 1, 1), None, False),
+        (date, '2017-01-01', date(2017, 1, 1), None, True),
+        (date, '2017-01-01T12:13:14.567', ValidationError, 'type=date_from_datetime_inexact', False),
+        (date, '2017-01-01T12:13:14.567', ValidationError, 'type=date_parsing', True),
+        (date, '2017-01-01T00:00:00', date(2017, 1, 1), None, False),
+        (date, '2017-01-01T00:00:00', ValidationError, 'type=date_parsing', True),
+        (datetime, '2017-01-01T12:13:14.567', datetime(2017, 1, 1, 12, 13, 14, 567_000), None, False),
+        (datetime, '2017-01-01T12:13:14.567', datetime(2017, 1, 1, 12, 13, 14, 567_000), None, True),
+    ],
+    ids=repr,
+)
+def test_validate_strings(field_type, input_value, expected, raises_match, strict):
+    if raises_match is not None:
+        print(TypeAdapter(field_type).core_schema)
+        with pytest.raises(expected, match=raises_match):
+            TypeAdapter(field_type).validate_strings(input_value, strict=strict)
+    else:
+        TypeAdapter(field_type).validate_strings(input_value, strict=strict) == expected
+
+
+@pytest.mark.parametrize('strict', [True, False])
+def test_validate_strings_dict(strict):
+    assert TypeAdapter(Dict[int, date]).validate_strings({'1': '2017-01-01', '2': '2017-01-02'}, strict=strict) == {
+        1: date(2017, 1, 1),
+        2: date(2017, 1, 2),
+    }

